### PR TITLE
Fix typical acceptance sampler with correct recovered token ids

### PR DESCRIPTION
**Pull request by jiqing-feng**
Fix typical acceptance sampler with correct recovered token ids

### CODE REVIEW

**Comment by jiqing-feng** [comment]
Hi @LiuXiaoxuanPKU , as we discussed in #8320 . This PR benefits the typical acceptance sampler with 20%+ speed-up (target model: llama-7b;  draft_model: llama-68m).

**Comment by jiqing-feng** [comment]
Hi @sroy745 @cadedaniel . Do you mind reviewing this PR? It is an optimization for increasing the acceptance rate without precision loss because we will always select 1 token from the target model.

**Comment by LiuXiaoxuanPKU** [comment]
Sorry for the delay, I will review it by tmr.

**Review by LiuXiaoxuanPKU — COMMENTED** [review]
On the high level, are you saying the old way of _replacement_token_ids is 20% slower than the current way of _get_recovered_token_ids even if mathematically, the two implementations are the same when k=1?

**Comment by jiqing-feng** [comment]
> On the high level, are you saying the old way of _replacement_token_ids is 20% slower than the current way of _get_recovered_token_ids even if mathematically, the two implementations are the same when k=1?

Yes, exactly. The optimization only works when k > 1.

File: vllm/model_executor/layers/typical_acceptance_sampler.py [hunk 1]
```diff
@@ -173,13 +167,9 @@ def _replacement_token_ids(self, target_probs):
...
+            A tensor of shape (batch_size, k) with the recovered token
+            ids which are selected from target probs.
         """
-        max_indices = torch.argmax(target_probs[:, 0, :], dim=1)
-        output = -torch.ones((target_probs.shape[0], target_probs.shape[1]),
-                             dtype=self.token_id_dtype,
-                             device=target_probs.device)
-        output[:, 0] = max_indices
-        return output
+        max_indices = torch.argmax(target_probs, dim=-1)
```
[the remark this answers is not in the document]
Comment (jiqing-feng) [reply vllm/model_executor/layers/typical_acceptance_sampler.py:173]: That's because we didn't check the recovered token
image

File: vllm/model_executor/layers/typical_acceptance_sampler.py [hunk 2]
```diff
@@ -173,13 +167,9 @@ def _replacement_token_ids(self, target_probs):
         Returns
         -------
         torch.Tensor
-            A tensor of shape (batch_size, k) with the replacement 
-            token IDs. Only the first column is set, and the rest of the
-            columns are filled with -1.
+            A tensor of shape (batch_size, k) with the recovered token
```
[the remark this answers is not in the document]
Comment (jiqing-feng) [reply vllm/model_executor/layers/typical_acceptance_sampler.py:170]: The optimization works only when k>1. If k=1, it will remain the same as previous.

File: vllm/model_executor/layers/typical_acceptance_sampler.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (sroy745) [reply vllm/model_executor/layers/typical_acceptance_sampler.py:173]: Thanks. Can you please update the test to verify this?

**Review by sroy745 — COMMENTED** [review]
Left one comment to update the unit test. Otherwise LGTM

**Comment by LiuXiaoxuanPKU** [comment]
Hi, overall this looks good to me. Could you
1. make this PR ready so that we can check it passes CI for other tests.
2. Could you check medusa's implementation and see how they handle the recover tokens? I'm a bit concerned if this change will affect the model accuracy, but the intuition makes sense to me and the speedup seems good.

**Review by LiuXiaoxuanPKU — APPROVED** [review]

**Comment by jiqing-feng** [comment]
> Hi, overall this looks good to me. Could you
> 
> 1. make this PR ready so that we can check it passes CI for other tests.
> 2. Could you check medusa's implementation and see how they handle the recover tokens? I'm a bit concerned if this change will affect the model accuracy, but the intuition makes sense to me and the speedup seems good.

1. This PR has been rebased on the latest change and all tests have been cleaned.
2. I have tested medusa by following the tests [here](https://github.com/vllm-project/vllm/blob/main/tests/spec_decode/e2e/test_medusa_correctness.py) and there is no accuracy regression. Please let me know if you want me to test more Medusa models.

BTW, I have successfully passed the `tests/spec_decode/e2e/test_medusa_correctness.py` and `tests/samplers/test_typical_acceptance_sampler.py` locally. Please let me know if any other tests are required.

**Comment by LiuXiaoxuanPKU** [comment]
> > Hi, overall this looks good to me. Could you
> > 
> > 1. make this PR ready so that we can check it passes CI for other tests.
> > 2. Could you check medusa's implementation and see how they handle the recover tokens? I'm a bit concerned if this change will affect the model accuracy, but the intuition makes sense to me and the speedup seems good.
> 
> 1. This PR has been rebased on the latest change and all tests have been cleaned.
> 2. I have tested medusa by following the tests [here](https://github.com/vllm-project/vllm/blob/main/tests/spec_decode/e2e/test_medusa_correctness.py) and there is no accuracy regression. Please let me know if you want me to test more Medusa models.
> 
> BTW, I have successfully passed the `tests/spec_decode/e2e/test_medusa_correctness.py` and `tests/samplers/test_typical_acceptance_sampler.py` locally. Please let me know if any other tests are required.

Good to go. Thanks for the work.
